### PR TITLE
SCT-115 Enable Task to accept externally generated task id

### LINKED ISSUE #115: Enable Task to accept externally generated task id
In order to allow for the time difference between when a task is requested and the infrastructure actually launches it, we'll need to be able to map the infrastructure's execution id with the task execution id.  This story is to allow that mapping to occur by having the launcher do the creation of the task execution record within the task repository.

### PR DESCRIPTION
**Pull request by cppwfs**
resolves #115

### CODE REVIEW

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/TaskRepository.java [hunk 1]
```diff
@@ -68,4 +68,24 @@ TaskExecution completeTaskExecution(long executionId, Integer exitCode, Date end
 	TaskExecution createTaskExecution(String taskName,
 			Date startTime,List<String> arguments);
 
+	/**
+	 * Notifies the repository that an empty taskExecution needs to be created. TaskName, startTime are set to null
```
Comment (mminella) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/TaskRepository.java]: Can we change this to something like "Creates an empty `TaskExecution` with just an id provided.  This is intended to be utilized in systems where the request of launching a task is separate from the actual start of a task (the underlying system may need to deploy the task prior to launching, etc)."?

File: spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/dao/MapTaskExecutionDao.java [hunk 2]
```diff
@@ -62,7 +62,22 @@ public TaskExecution createTaskExecution(String taskName,
...
+		taskExecution.setArguments(arguments);
+
+		return taskExecution;
+	}
+
+	@Override
 	public void completeTaskExecution(long executionId, Integer exitCode, Date endTime, String exitMessage, String errorMessage) {
+		if(!this.taskExecutions.containsKey(executionId)) {
+			throw new IllegalStateException("Invalid TaskExecution, ID " + executionId + " not found.");
+		}
```
Comment (mminella) [inline spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/dao/MapTaskExecutionDao.java]: Should we add the completed check here as well?

File: spring-cloud-task-core/src/test/java/org/springframework/cloud/task/repository/dao/MapTaskExecutionDaoTests.java [hunk 3]
```diff
@@ -16,43 +16,86 @@
...
+				taskExecutionMap.get(expectedTaskExecution.getExecutionId()));
+	}
+
+	@Test(expected = IllegalStateException.class)
+	public void completeTaskExecutionWithNoCreate() {
+		TaskExecution expectedTaskExecution = TestVerifierUtils.createSampleTaskExecutionNoArg();
+		this.dao.completeTaskExecution(expectedTaskExecution.getExecutionId(),
+				expectedTaskExecution.getExitCode(), expectedTaskExecution.getEndTime(),
+				expectedTaskExecution.getExitMessage());
+	}
```
Comment (mminella) [inline spring-cloud-task-core/src/test/java/org/springframework/cloud/task/repository/dao/MapTaskExecutionDaoTests.java]: Can we get a test that tests completing a task that is already completed?

File: spring-cloud-task-integration-tests/src/test/java/org/springframework/cloud/task/executionid/TaskStartTests.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+		myapp.setEnvironment(environment);
+		return myapp;
+	}
+
+	private boolean tableExists() throws SQLException {
+		boolean result;
+		try (
+				Connection conn = dataSource.getConnection();
+				ResultSet res = conn.getMetaData().getTables(null, null, "TASK_EXECUTION",
+						new String[]{"TABLE"})) {
```
Comment (mminella) [inline spring-cloud-task-integration-tests/src/test/java/org/springframework/cloud/task/executionid/TaskStartTests.java]: Indentation of this block is off...

**Comment by cppwfs** [comment]
Sounds good to me.

**Comment by mminella** [comment]
Merge as 441bbfe49240c1773f3b5d99f3488defb17318eb

**Pull request closed without merging**